### ISSUE TITLE
Make it easier to write a build.yaml

### ISSUE THREAD

**Issue by natebosch** [dart-lang/build#885]
There are a lot of sharp edges today.

Some initial ideas:

- Split 'targets' and 'builders' into separate files. This would be breaking so rollout would be tough - but it could be worth it in the long run.
  - Most consumers will only need a 'targets' section so can have more focused documentation and examples
  - Reduces indentation by a level
  - Targets have a 'builders' key - and it's easy to get this confused with the current top-level 'builders' key
  - Con: Less future proof if we need to add another top-level concept in addition to 'targets' and 'builders'
- Improve error messages. Today you just get a short error and a stack trace.
- Add a validation tool?

**Comment by natebosch** [comment]
Any validation should check for mixing `builders` section with `--config` usage. In #882 we'll not support changing the generated entrypoint based on that flag for now.

**Comment by matanlurey** [comment]
@natebosch:

> Split 'targets' and 'builders' into separate files.

How would you envision this working? A `build.yaml` and a `builders.yaml`?

> * Most consumers will only need a 'targets' section so can have more focused documentation and examples

+1

> Improve error messages. Today you just get a short error and a stack trace.

+1. Part of this will just be trial and error (and bugs filed).

Also +1 validation tool, and eventually an analyzer plugin.

**Comment by natebosch** [comment]
> How would you envision this working? A `build.yaml` and a `builders.yaml`?

Yeah, I was thinking `build.yaml` and `published_builders.yaml` to avoid confusion. Jake pointed out though that this does limit future-proof extensions, like having some `build.yaml` config for custom server handlers etc.

**Comment by matanlurey** [comment]
True. I think we can probably keep one file if the validation/docs are good enough.

**Comment by davidmorgan** [comment]
Closing as stale.
